presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -0,0 +1,232 @@
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator planNodeIdAllocator;
+        private final PlanVariableAllocator planVariableAllocator;
+        private final Map<VariableReferenceExpression, VariableReferenceExpression> keyToRandomKeyMap;
+
+        private Rewriter(Session session,
+                FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, PlanVariableAllocator planVariableAllocator)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.planNodeIdAllocator = requireNonNull(planNodeIdAllocator, "planNodeIdAllocator is null");
+            this.planVariableAllocator = requireNonNull(planVariableAllocator, "planVariableAllocator is null");
+            this.keyToRandomKeyMap = new HashMap<>();
+        }
+
+        private static boolean isSupportedType(VariableReferenceExpression variable)
+        {
+            return Stream.of(VARCHAR, BIGINT, DATE).anyMatch(x -> x.equals(variable.getType()));
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Void> context)
+        {
+            // In this optimization, we add the randomized key to output so it can be reused in case the same key is used in outer join in later stage.
+            // However this can change the input for a cross join (if the input of the cross join gets optimized). For cross join, it does not allow mismatch
+            // of input and output. Need to reassign output in case input of the cross join changes.
+            if (joinNode.isCrossJoin()) {
+                PlanNode rewrittenLeft = context.rewrite(joinNode.getLeft(), context.get());
+                PlanNode rewrittenRight = context.rewrite(joinNode.getRight(), context.get());
+                Set<VariableReferenceExpression> inputVariables = ImmutableSet.<VariableReferenceExpression>builder()
+                        .addAll(rewrittenLeft.getOutputVariables())
+                        .addAll(rewrittenRight.getOutputVariables())
+                        .build();
+                checkState(inputVariables.containsAll(joinNode.getOutputVariables()));
+                return new JoinNode(
+                        joinNode.getSourceLocation(),
+                        joinNode.getId(),
+                        joinNode.getStatsEquivalentPlanNode(),
+                        joinNode.getType(),
+                        rewrittenLeft,
+                        rewrittenRight,
+                        joinNode.getCriteria(),
+                        inputVariables.stream().collect(Collectors.toList()),
```

Comment:
Use `toImmutableList()` here and other places below

Reply:
Done